feat: introspection SDL encoder

lrlna opened the pull request:
## Description

As a first step of working with an Introspection result and creating an SDL, I am introducing a encoder that produces the correct output as part of a code generation step.

As part of this work, I am also aiming to establish a data flow of how an SDL will be produced, and so far it's as follows:
1. Query an endpoint. (work done in `crates/rover-client/introspect.rs`)
2. Receive an introspection result. (work done in `crates/rover-client/introspect.rs`)
3. Because `graphql_client` already returns a parsed result (i.e. we have some types we can work with), we can create an intermediary representation of schema that will be used to do SDL code generation. (a very _smol_ amount of work for this is done in `crates/rover-client/schema.rs).
4. `crates/rover-client/schema.rs` traverses the `types` and `directives` received from Introspection Result. It then uses teh SDL encoder to create an SDL string. 
4. SDL encoding is done using the `crates/sdl-encoder` which is introduced by this PR. This is responsible to serialising the output.

## Running this PR
Please run introspection on your favourite endpoints by checking out this branch and running:
```console
cargo run graph introspect http/to/endpoint
```

You may also wish to run the docs on the SDL encoder:
```console
cd crates/sdl-encoder && cargo doc --open
```

## About `sdl-encoder`
Design for this was influenced by [`wasm-encoder`](https://github.com/bytecodealliance/wasm-tools/tree/main/crates/wasm-encoder). This tool is doing a similar job of creating specific pieces of the final formatted output, and I find the library to be a well executed example to follow. 

The encoder is meant to be used in `crates/rover-client/schema.rs` as we build out the functionality of an intermediate schema, and work with the parsed input provided by IntrospectionResult.

*Most* GraphQL Introspection [response types](https://spec.graphql.org/draft/#sec-Schema-Introspection) are represented with this encoder. The exception is the `__InputObject` using the prescribed by the spec `__InputValue` type. As we do not have an AST to visit to encode the schema, all types that have fields currently use `__Field` type as their input. 

The encoder is written in a way for us to use and expand in the future to encode schemas that are not just part of Introspection response. It will require some tweaking, such as allowing Types to have other directives other than deprecated, but this can be added in a very straightforward manner when the time comes.

closes #180
closes #171

lrlna committed:
feat: create an SDL encoder

This introduces the concept of a SDL encoder/serialiser that will be
used when doing a code generation pass on an Introspection Result.

lrlna committed:
sdl-encoder: create a Field struct

Each type and schem defs can have a field. A field is a key value pair
describing a field and its type, which can be nullable, or a vector,
and can have a default. This commit also introduces a Display
implementation for Field and applicable tests.

lrlna commented on crates/sdl-encoder/src/lib.rs:
```diff
@@ -0,0 +1,373 @@
+                VecState::None => format!(" = {}{}", default, null),
+                VecState::Nullable => format!(" = [{}{}]", default, null),
+                VecState::NonNullable => format!(" = [{}{}]!", default, null),
+            }
```
I am currently encoding `default` values based on the shape of the original type, so let's say the field looks like:
```js
key: ValueEnum!
```
The default value will look like:
```js
key: ValueEnum! = EnumVersionOne!
```
If the field looks like this:
```js
key: [ValueEnum!]!
```
The default would look like this:
```js
key: [ValueEnum!]! = [EnumVersionOne!]!
```

Does this seem correct? Should I be making any adjustments?

martijnwalraven replied:
I hope I understand the question correctly, but fields by themselves don't have default values. Only field arguments do (and variables, but that isn't relevant for schemas).

As for how these default values are represented, you may not want to use a `string` for this but rely on a more structured way of representing values. In Apollo iOS we use [an enum](https://github.com/apollographql/apollo-ios/blob/main/Sources/ApolloCodegenLib/Frontend/GraphQLValue.swift) for this, which is similar to [what `graphql-js` does in its AST](https://github.com/graphql/graphql-js/blob/735975d4904c5d9cc93d0c0a02af181e8f02dfba/src/language/ast.d.ts#L309-L318).

lrlna replied:
I am sorry, I did mean `value`, or field argument in a given field. I was looking at a few SDLs and wanted to find out what variations there are for a way to display a default. I found things like this:

```js
input DirectorySortInput {
  fields: [DirectoryFieldsEnum]
  order: [SortOrderEnum] = [ASC] // this is a default assignment, correct?
}
```
and
```js
type ImageSharp implements Node {
    srcSetBreakpoints: [Int] = [] // so are these three?
    toFormat: ImageFormat = NO_CHANGE
    toFormatBase64: ImageFormat = NO_CHANGE
}
```

And I am wondering what other ways there could be to set a default, in case my assumption is incorrect.

As for default value representation, everything is going to be a `String` as this is a serialiser and all it concerns itself with is how to stringify the data.

martijnwalraven replied:
The second example isn't actually valid, because fields on object types and interfaces can't have default values. Only fields on input objects or field arguments can.

So this is an additional place where default values might come in:
```
type Product {
  weight(unit: WeightUnit = KILOS): Float
}
```

martijnwalraven replied:
For that reason, `graphql-js` differentiates between `GraphQLField` (which doesn't have a default value) and `GraphQLInputField` (which does).

lrlna replied:
Ah ok! gotcha. I have not seen a default for a field argument example yet, so this is useful.

lrlna replied:
Given the example you used what are the different components in the statement:
```js
type Product {
  weight(unit: WeightUnit = KILOS): Float
}
```
i.e. what would you call  these components: `unit`, `WeightUnit` and `KILOS`?

lrlna commented on crates/sdl-encoder/src/lib.rs:
```diff
@@ -0,0 +1,373 @@
+        // TODO(@lrlna): double check with folks if it's a valid SDL if the last
+        // field in a type has a comma. If not, we can move the 'comma logic' to
+        // TypeDef/SchemaDef Display implementations.
+        write!(f, "  {}: {}{},", &self.name, type_, default)
```
Can the last field in a type definition have a comma, like so:
```js
type Query {
  key: String,
  key2: Int,
  key3: String, // <--- this comma here valid?
}
```

abernix replied:
@lrlna It can have a comma there, yes, but also _all the commas are optional_.

lrlna replied:
eh?????? 

I'll drop the comma then, easier for me \o/

martijnwalraven replied:
It's pretty weird, but commas are treated as whitespace in the GraphQL grammar, so you can insert them anywhere you want. Even `type,Query,{key:,String,},}` would be valid…

martijnwalraven commented on crates/sdl-encoder/src/lib.rs:
```diff
@@ -0,0 +1,373 @@
+pub struct Field {
+    description: Option<String>,
+    name: String,
+    type_: String,
```
You probably don't want a type represented as a `string` here, but in a more structured way, like in [`graphql-js`'s `GraphQLType` hierarchy](https://github.com/graphql/graphql-js/blob/735975d4904c5d9cc93d0c0a02af181e8f02dfba/src/type/definition.d.ts#L36).

martijnwalraven commented on crates/sdl-encoder/src/lib.rs:
```diff
@@ -0,0 +1,373 @@
+
+/// Define whether a Field is a vector.
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum VecState {
```
This won't work for nested lists (e.g. `[[String]]`), and it also won't allow you to differentiate between nullable and non-nullable element types (e.g. `[String]` vs `[String!]`). That might be a good reason to use a more structured representation (probably an enum, like we [do in Apollo iOS](https://github.com/apollographql/apollo-ios/blob/cd9a77bac40c61f547a6c942bca2fcd74114d0d5/Sources/ApolloCodegenLib/Frontend/GraphQLType.swift#L4)).

lrlna replied:
Ah yes, this does not account for `[[String]]`! the `[String]` and `[String!]` is accounted for within the Field type under the `nullable` property. I'll think about a way to accommodate lists though.

martijnwalraven replied:
Note that there is a difference between `[String!]` (nullable list with a non-nullable element type)  and `[String]!` (non-nullable list with a nullable element type) though (and `[String!]!`, `[[String!]!]`, and `[[String]!]`, etc.). I really think this requires a recursive type definition to get it right.

lrlna replied:
yes, of course the parsed introspection result will have to be read recursively. But that's not part of the serialiser's job -- that will be handled in a different part of an application. This module literally only cares about creating the string.

martijnwalraven replied:
But what I mean is that in order to generate the correct string you need to be able to differentiate between these different types. It isn't enough to know that a field is nullable, you need to know if the element type of a list (if it is a list) is nullable, or even keep track of nullability separately at different levels in the case of nested lists. I don't think this has to complicate the code too much, this is [how we generate strings recursively for type references in Apollo iOS](https://github.com/apollographql/apollo-ios/blob/cd9a77bac40c61f547a6c942bca2fcd74114d0d5/Sources/ApolloCodegenLib/Frontend/GraphQLType.swift#L9).

lrlna replied:
Ah I see what you mean! I think I have a good way to fix this.

lrlna committed:
feat(sdl-encoder): Fields can have FieldType

FieldType encompasses the idea that type name can be infinitely
recurssed in a Nullable or a NonNullable List. To accommodate that we
introduce a new Enum that references itself in a FieldType::List
variant.

lrlna committed:
feat(sdl-encoder): SDL can have an input object

This commit adds an ability to add input objects to the overarching
SDL schema.

lrlna committed:
feat(introspection): parse introspection schema

This commit creates a new function on the introspetion schema's impl to
parse the incoming introspection result and create an SDL with the
sdl-encoder.

lrlna committed:
feat(introspection): parse introspection schema

This commit creates a new function on the introspetion schema's impl to
parse the incoming introspection result and create an SDL with the
sdl-encoder.

lrlna committed:
feat(sdl-encoder): add enum definition

This allows for encoding of enum types that will look like this:
```js
enum VeryGoodCats {
  NORI
  CHASHU
}
```

lrlna committed:
feat(sdl-encoder): add scalar encoding

Creates encoding for scalar type: `scalar NameOfScalar`.

lrlna committed:
chore(sdl-encoder): refactor for easier reading

1. Iterate over the first level of NON_NULL
Because the ResponseData creates its own types and does not create
types we can recurse over in an introspection response, we have to
manually unwrap Introspection's `ofType`. This commit unwraps the
initial NON_NULL.

2. Refer to enum's variants as 'values'

3. clarify comment for object type/interface arguments

EverlastingBugstopper commented on crates/sdl-encoder/src/lib.rs:
```diff
@@ -315,20 +403,39 @@ mod tests {
+        // enum definition
+        let mut enum_ = EnumDef::new("VeryGoodCats".to_string());
+        enum_.variant("NORI".to_string());
+        enum_.variant("CHASHU".to_string());
```
😼

lrlna commented on crates/rover-client/src/introspection/schema.rs:
```diff
@@ -0,0 +1,295 @@
+    // TypeRefOfTypeOfTypeOfType
+    // TypeRefOfTypeOfTypeOfTypeOfType
+    // TypeRefOfTypeOfTypeOfTypeOfTypeOfType
+    fn encode_field(field: FullTypeFields) -> Field {
```
This is an incredibly lengthy function that encodes individual fields, such as: `name: NameType`, `versions: [Version]!`, `treats: [[TreatObject]!]!`. Currently, this will only handle lists up 2, meaning `[[MyType]]`. Given some testing I've done, I will have to expand this manual process for `[[[[MyType]]]]`, which means this function will grow a lot. I would like y'all to take a look at at this state first, before it will become entirely unmanageable to review.

**Help**: I could definitely use another pair of eyes on this, as I am not 100% certain `NON_NULL` and `LIST` nullability setting is correct. So for a (pseudo) introspection result that has:
```
NON NULL
  LIST
    NON NULL
      LIST
         NON NULL
            SCALAR
```
, I want to make sure I applied the nullability to the correct types. I keep re-reviewing this and getting lost because my brain was not meant for manual recursion 🙈

The nullable portion is set in `sdl-encoder`'s `FieldType`. So [here](https://github.com/apollographql/rover/pull/283/files#diff-0a67f645a01689ab4ea28b13a07ab090ec5869d658ae691cb84a8e30ed6bbcd0R81) `let ty = FieldType::new_type(type_ref.name.unwrap(), true, None);`, the `true` refers to the fact that this type `is_nullable`, or can in fact be nullable, meaning I don't append the `!` in the SDL. Hope that makes sense!

martijnwalraven replied:
I left a comment below as well, but I think following `graphql-js` here and making `NonNull` a type wrapper similar to `List` (as opposed to a boolean argument) would help simplify this.

martijnwalraven replied:
Also see this [`GraphQLType` enum definition in Apollo iOS](https://github.com/apollographql/apollo-ios/blob/d1369d43cd38896547278f147a8654146e864448/Sources/ApolloCodegenLib/Frontend/GraphQLType.swift#L3-L19), which shows how that allows for easy recursive printing of type references.

lrlna replied:
This function is now recursive, so I'll resolve this set of comments. The `NonNull` type wrapper comment is kept open over at the `sdl-encoder`.

lrlna commented on crates/rover-client/src/introspection/schema.rs:
```diff
@@ -0,0 +1,295 @@
+                                field_def.description(field.description);
+                                field_def
+                            }
+                            LIST => panic!("Cannot nest more than 2 lists."),
```
I realise it's a *very giant* faux pas to have the encoding step panic/error. There are a few places here where we panic, specifically with LIST. I will be expanding these lists to be able to nest up to 4, but I will have to panic after that until we write our own parser.

lrlna commented on crates/rover-client/src/introspection/schema.rs:
```diff
@@ -0,0 +1,295 @@
+                    ty => panic!("Unknown type: {:?}", ty),
+                }
+            }
+            Other(ty) => panic!("Unknown type: {}", ty),
```
**Help**: What's this `Other` type that I keep getting back as part of the results? Are there other types that also need encoding? Would someone be able to point me to the documentation on these, or ideally the spec?

martijnwalraven replied:
I have no idea where `Other` comes from, that's not part of the spec so this may be an artifact of how `graphql-client` works?

lrlna replied:
If you don't know of any other types other than SCALAR,  OBJECT,  INTERFACE, UNION, ENUM, INPUT_OBJECT, LIST and NON_NULL, I take it would be an artefact of graphql_client. I have not been able to cause a panic coming from here, so I am hoping we will never hit this case.

martijnwalraven replied:
That seems right, both according to the spec and looking at the introspection code in `graphql-js`. Maybe `graphql-client` just treats every enum as open ended? We do something similar in Apollo iOS to allow for schema evolution.

lrlna commented on crates/rover-client/src/introspection/schema.rs:
```diff
@@ -0,0 +1,295 @@
+impl TryFrom<Introspection> for Schema {
+    type Error = &'static str;
+
+    fn try_from(src: Introspection) -> Result<Self, Self::Error> {
```
A note on this, and the lack of a constructor method on Schema.

Because the Introspection response may or may not have a malformed result, we need to make sure construction of `Schema` is allowed to fail. None of the usual `new`, `with_`, and `from_` constructor methods are allowed to fail, so our by-the-books way of constructing a Schema is using a `TryFrom` trait implementation.

EverlastingBugstopper replied:
TIL `new` is not supposed to be fallible!! I... might need to update a library or two 😆

lrlna commented on crates/rover-client/src/introspection/schema.rs:
```diff
@@ -0,0 +1,295 @@
+#[cfg(test)]
+mod tests {
+    #[test]
+    fn it_build_simple_schema() {}
```
This is empty so far, unfortunately. I am thinking of a good way to be able to run 10 or so overall tests for overall schema encoding. The problem is that I need the actual `Introspection` type that I think I can literally only get by using and making queries to the graphql_client, which is really quite bulky. If I can't come up with anything smarter, that's what I'll end up doing, I think.

**Help**: If you can think of another clever way, please let me know!

martijnwalraven replied:
Couldn't you manually execute introspection queries  and store the results in files?
(With the old `apollo` CLI, you could get those by running something like `apollo schema:download --endpoint=http://localhost:8080/graphql schema.json`.) Or am I misunderstanding the issue?

lrlna replied:
What you're suggesting is not possible as `Schema` is currently created for an `IntrospectionResult`. `.json` file will not work. A `.json` file would  have worked if `IntrospectionResult` had a `Serialize` implementation, but this implementation doesn't exist and the following code does not work:
```rust
        let file = File::open("src/introspection/introspection.json")
            .expect("File should be able to be opened.");
        let result: IntrospectionResult =
            serde_json::from_reader(file).expect("File is not a proper JSON.");
```

martijnwalraven replied:
I don't understand. Since executing an introspection query against a server would return the same JSON contents, and we are able to load those into an `IntrospectionResult`, why is this different?

lrlna commented on crates/sdl-encoder/src/lib.rs:
```diff
@@ -0,0 +1,292 @@
+//! SDL Encoder provides methods to serialise a GraphQL Schema.
```
This crate is what does the actual SDL creation. It's divided into smaller `mod`s based on various types we can encounter in a Schema, plus two convenience impls: `Field` and `FieldType`. Overall, we are trying to do the simplest thing possible -- create a buffer, keep adding to it as we iterate over introspection result, and at the end call `.finish` to be able to get the final string. All individual types and helpers have a Display implementation which make this whole thing really straightforward.

lrlna reviewed (commented):
I'd like to send this over for a first pass of reviews. I've left a few comments about certain items. If there is anything that feels like you'd like to talk over zoom, please let me know. 

You can run this work with `cargo run graph introspect /url/of/graph/to/introspect`. For example this one `cargo run graph introspect https://api.react-finland.fi/graphql` gives you a nice (non `panic`ed) SDL. As noted in [one of the comments](), this only so far does ofType encoding for up-to `[[Type!]!]!`.

A couple of things here -- not all things have encodings quite yet. Things that I am aware so far that I need to encode: `directives`, arguments (i.e. `(includeDeprecated: Boolean = false)` in `fields(includeDeprecated: Boolean = false): [__Field!]`), `union`, `interface` and being able to provide default arguments. `Union`, `Interface`, and (_I think_) `directives` are incredibly straightforward. I have to think a tad more about arguments and defaults though. If  this list is missing something else, **please let me know**.

martijnwalraven commented on crates/rover-client/src/introspection/schema.rs:
```diff
@@ -0,0 +1,295 @@
+        }
+    }
+
+    // NOTE(lrlna): We manually unroll our ofTypes instead of recursing here
```
I take it [this](https://github.com/graphql-rust/graphql-client/issues/312) is the underlying issue in `graphql-client`? 

I really wish we could replace all this code by a recursive function, similar to the [`getType`](https://github.com/graphql/graphql-js/blob/main/src/utilities/buildClientSchema.js#L124-L141) function used to get a type from an introspection result in `graphql-js`. Would it be an option to use an untyped JSON result for now (e.g. `serde_json::Value`) so we don't have to unroll this manually?

martijnwalraven replied:
Or alternatively, would it be possible to define a `FullType` trait to represent the fragment and have the structs that represent the places it is spread implement that?

lrlna replied:
So using `serde_json::Value` would mean we would have to write our own code generation on top of already completed code generation done by graphql_client. I feel like at that point, I'd rather just write our own parser (which is the intention, but it's further down the line).

I am not entirely sure what you mean by your second comment, and perhaps what you suggested is what I ended up implementing. You can take a look at the `OfType` trait and [its helper macro here](https://github.com/apollographql/rover/pull/283/files#diff-3495ffcb38fab94efec3c26b550005475c0e9b80718b7d3f4d0a0bdc39264154R52-R131).

martijnwalraven commented on crates/sdl-encoder/src/enum_def.rs:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    /// Set the EnumDef's variants.
+    pub fn variant(&mut self, variant: String) {
```
I believe both the spec and `graphql-js` refer to these as enum `values`.

lrlna replied:
oh, i didn't know, thanks! I just went with the Rust's term.

martijnwalraven commented on crates/sdl-encoder/src/field.rs:
```diff
@@ -0,0 +1,114 @@
+#[derive(Debug, PartialEq, Clone)]
+pub struct Field {
+    description: Option<String>,
+    //TODO(@lrlna): fields for input objects can also take arguments. This
```
To clarify, it's the other way around: fields on object types (and interfaces) can take arguments, but fields on input objects do not. Another difference is that input object fields can have default values. For these reasons, it might make sense to follow `graphql-js` and differentiate between `Field` and `InputField`.

lrlna replied:
ok! thanks! I clarified the comment.

martijnwalraven replied:
Oh, another reason to differentiate between `Field` and `InputField` is that the type of an input object has to be an input type (meaning it can't be an object type/interface/union). You don't need this for printing out introspection results as SDL, but we may want to encode those constraints when building a more complete GraphQL library.

(In `graphql-js` this is represented as a [`GraphQLInputType` discriminated union](https://github.com/graphql/graphql-js/blob/main/src/type/definition.d.ts#L87), but there may be a better way to encode this in Rust.)

lrlna replied:
For now I'll keep `InputObject` to have a regular `Field` for now. Once we have an AST we can visit when encoding schemas, we can bring in an `InputField` as that would be a lot easier. I made a note specifically for this in `InputObject` struct definition.

martijnwalraven commented on crates/sdl-encoder/src/field.rs:
```diff
@@ -0,0 +1,114 @@
+            // are always on the same level and are indented by 2 spaces.
+            writeln!(f, "  \"\"\"\n  {}\n  \"\"\"", description)?;
+        }
+        // TODO(@lrlna): double check with folks if it's a valid SDL if the last
```
Commas are valid everywhere, and treated as whitespace. That means there is no reason to use commas either (as long as there is another whitespace separator between tokens).

martijnwalraven commented on crates/sdl-encoder/src/field_type.rs:
```diff
@@ -0,0 +1,65 @@
+
+/// Define Field Type.
+#[derive(Debug, PartialEq, Clone)]
+pub enum FieldType {
```
You may want to call this `Type` instead of `FieldType` because the same structure is used for types of arguments.

(In `graphql-js`, this would correspond to `GraphQLType`, and there is a whole hierarchy of types and type comparison functions that we may want to support if we extend this beyond the current introspection use case.)

lrlna replied:
Hmmm yea that's a good point, thanks!

martijnwalraven commented on crates/sdl-encoder/src/field_type.rs:
```diff
@@ -0,0 +1,65 @@
+        /// List field type.
+        ty: Box<FieldType>,
+        /// Nullable list.
+        is_nullable: bool,
```
I think it may make sense to follow `graphql-js` here and make `NonNull` a type wrapper similar to `List`.

lrlna replied:
That's an interesting idea, let me think of a good way of making it work here.

lrlna replied:
`NonNull` type worked really nicely, thanks!

martijnwalraven reviewed (commented):
I think this is still missing support for interfaces, unions, field arguments, and directives?

lrlna committed:
feat(sdl-encoder): add encoding for Directives.

This commit adds Directive encodings. So far no arguments can be
passed, but we are able to accept description and locations. The output
is as follows:
```js
directive @cacheControl on FIELD_DEFINITION | INPUT_FIELD_DEFINITION
```

lrlna committed:
chore: ofType can be recursive, as a treat

Since all of the OfType types are unique in Introspection Response,
**and** we want to be able to recurse over many nested versions of
them, let's create an OfType Trait that can be implemented for all
underlying OfType types. To make the trait implementations a bit
easier, we are also creating a macro.

lrlna committed:
feat(sdl-encoder): add encoding for Unions

Unions are now encoded into appropriate SDL format: `union Cat = NORI | CHASHU`, for example.

lrlna committed:
feat(sdl-encoder): encoder Interface types

This commit creates an Interface type, as well as allow to specify if a
specific type implements an interface. This means we are able to create:
```js
interface Meal {
  main: String
  """
  Cat's post meal snack.
  """
  snack: [String!]!
  """
  Does cat get a pat after meal?
  """
  pats: Boolean
}
```
, as well as:
```js
type Dinner implements Meal {
}
```

lrlna committed:
feat(sdl-encoder): allow for deprecated fields

This commit adds a `deprecated` and `deprecation_reason` to the Field
struct. This lets each field to be deprecated and encoded with, for
example:
```js
cat: [SpaceProgram] @deprecated(reason: "Cats are no longer sent to
space.")
```

lrlna committed:
feat(sdl-encoder): add NonNull FieldType

adds NonNull as an enum variant for a FieldType. This makes it easier
to recurse and create nested non-nullable field values.

lrlna committed:
feat(sdl-encoder): fields can have arguments

Allows for various fields to have arguments. To make it easier to
iterate over the Introspection Response, we do not differentiate
between the various `__TypeKind`s that are _able_ to have an argument,
so all `__TypeKind`s with fields will check for an `args` field.

lrlna committed:
feat(introspection): exclude graphqltypes from SDL

lrlna committed:
feat(sdl-encoder): directives can have arguments

lrlna committed:
chore(sdl-encoder): seperate out possible values

FieldType was previously the only possible value a type could have.
This commit introduces EnumValue that is applicable for EnumTypes,
as well as renames FieldType to FieldValue to standardise Value naming.

lrlna committed:
feat(sdl-encoder): default values

This commit allows to set default values to Fields and InputValues. It
also renames FieldArguments to InputValues to align with the graphql
spec.

lrlna committed:
fix(sdl-encoder): multiline descriptions on "\n"

To align a bit closer on descriptions in Studio, we are now also
multilining descriptions only when a new line symbol is present in the
description.

lrlna committed:
chore(rust-docs): document sdl_encoder

This commit ensures that all structs and enums in sdl_encoder are
documented for docs.rs. It also cleans up struct and parameter names to
be as consistent as possible with GraphQL specification.

lrlna committed:
chore: align introspection with main branch

martijnwalraven commented on crates/rover-client/src/introspection/schema.rs:
```diff
@@ -0,0 +1,237 @@
+use serde::Deserialize;
+use std::convert::TryFrom;
+
+pub type Introspection = introspect::introspection_query::ResponseData;
```
Maybe `IntrospectionResult`?

martijnwalraven commented on crates/rover-client/src/introspection/schema.rs:
```diff
@@ -0,0 +1,237 @@
+    "__Directive",
+    "Boolean",
+    "String",
+    "Upload",
```
`Upload` is not a built-in type (it is part of the third-party [`graphql-upload` library](https://github.com/jaydenseric/graphql-upload)), so we probably don't want to skip encoding it to SDL.

martijnwalraven commented on crates/rover-client/src/introspection/schema.rs:
```diff
@@ -0,0 +1,237 @@
+pub type __TypeKind = introspect::introspection_query::__TypeKind;
+
+// Represents GraphQL types we will not be encoding to SDL.
+const GRAPHQL_NAMED_TYPES: [&str; 13] = [
```
Maybe we can find something more descriptive than `GRAPHQL_NAMED_TYPES`, or even expose a function that can be used as a filter predicate. `graphic-js` has a [`isDefinedType`](https://github.com/graphql/graphql-js/blob/main/src/utilities/printSchema.js#L49-L51) method that is used during schema printing to check whether a type is user-defined (the built-in types are filtered out by using `isSpecifiedScalarType` and `isIntrospectionType`).

lrlna replied:
Do you have a better name suggestion? I got `GRAPHQL_NAMED_TYPES` from [here](https://github.com/graphql/graphql-js/blob/main/src/utilities/printSchema.js#L56)

martijnwalraven replied:
`printFilteredSchema` allows you to pass in a `typeFilter`, which is used the filter named types. But `GRAPHQL_NAMED_TYPES` is specifically a list the of named types you want to filter out. `graphql-js` splits these up into introspection types ([`isIntrospectionType`](https://github.com/graphql/graphql-js/blob/540f33605ee047ab86e0d9666f4a60526d56ff0f/src/type/introspection.js#L531-L546)) and scalars that are included in the spec ([`isSpecifiedScalarType`](https://github.com/graphql/graphql-js/blob/540f33605ee047ab86e0d9666f4a60526d56ff0f/src/type/scalars.js#L272-L273)). It may make sense to follow that example and even use something like `isDefinedType` to check for types that are neither.

lrlna replied:
Ah, right. I think filters can be introduced at a later stage. For now it's easier (and more efficient) to have these as const. (The filter function that will then be introduced should also have these as consts, as these values are explicitly evaluated at compile time)

martijnwalraven commented on crates/rover-client/src/introspection/schema.rs:
```diff
@@ -0,0 +1,237 @@
+];
+
+// Represents GraphQL directives we will not be encoding to SDL.
+const GRAPHQL_DIRECTIVES: [&str; 3] = ["skip", "include", "deprecated"];
```
`graphql-js` calls these specified directives, and exposes an `isSpecifiedDirective` function that tests for them.

martijnwalraven commented on crates/rover-client/src/introspection/schema.rs:
```diff
@@ -0,0 +1,237 @@
+    }
+
+    fn encode_full_type(type_: SchemaTypes, sdl: &mut SDL) {
+        let ty = type_.full_type;
```
I assume we're using `ty` here because `type` is a reserved word? Is that a common abbreviation in Rust?

lrlna replied:
`type` is a reserved word, but I usually use `ty` as it's a common compiler abbreviation. Two examples to check out here:
- the rust compiler has a dedicated [ty module](https://github.com/rust-lang/rust/tree/master/compiler/rustc_middle/src/ty), and
- wasm-tools, specifically the wasm-encoder, uses [ty in most of their encodings](https://github.com/bytecodealliance/wasm-tools/blob/a1535a01ea425312258729a277a2fcee8e98aadb/crates/wasm-smith/src/encode.rs#L50-L81)

martijnwalraven commented on crates/rover-client/src/introspection/schema.rs:
```diff
@@ -0,0 +1,237 @@
+pub type Introspection = introspect::introspection_query::ResponseData;
+pub type SchemaTypes = introspect::introspection_query::IntrospectionQuerySchemaTypes;
+pub type SchemaDirectives = introspect::introspection_query::IntrospectionQuerySchemaDirectives;
+pub type FullTypeFields = introspect::introspection_query::FullTypeFields;
```
Why are these types plural if they represent a single object? Wouldn't `SchemaType`, `SchemaDirective`, and `FullTypeField` be more appropriate? Or perhaps `IntrospectionType`, `IntrospectionDirective`, and `IntrospectionField` (those are some of [the types used in `graphql-js` to represent the result of an introspection query](https://github.com/graphql/graphql-js/blob/main/src/utilities/getIntrospectionQuery.js#L153-L154))?

lrlna replied:
ah yea, thanks, these were just autocompleted from the generated `introspection-query` struct, and the names there are in plural.

martijnwalraven commented on crates/sdl-encoder/src/directive_def.rs:
```diff
@@ -0,0 +1,170 @@
+    args: Vec<InputValue>,
+    // Locations returns a List of __DirectiveLocation representing the valid
+    // locations this directive may be placed.
+    locations: Vec<String>,
```
We may not need this right now, but it may make sense to use a `DirectiveLocation` enum here instead of `String`.

lrlna replied:
Why would you want this as an enum? Is there other information about a directive we need to serialise?

martijnwalraven replied:
You'd want this as an enum if you're doing anything else than encoding it, like validating directive locations in a document. I also assume that would allow for a more compact representation and make it possible to store this as a set for easy membership checks for example.

lrlna replied:
This module should not be doing anything other than encoding. Any sort of validation, checks or referencing should be done in other parts of the system given an AST.

martijnwalraven commented on crates/sdl-encoder/src/field_value.rs:
````diff
@@ -0,0 +1,120 @@
+/// assert_eq!(non_null.to_string(), "[String]!");
+/// ```
+#[derive(Debug, PartialEq, Clone)]
+pub enum FieldValue {
````
Where does the name `FieldValue` come from? Isn't this a representation of any GraphQL type, that would also be used for the type of field and directive arguments? Maybe call this `Type` or `GraphQLType`, and refer to the `Type` variant as `NamedType`?

lrlna replied:
I find it confusing to call it a `GraphQLType`, even more so when it's just `Type`. There isn't anything in the spec about this particular bit of SDL, so I figured we'd call this something that's a bit more easily understood. 

`NamedType` sounds good! I'll use that instead.

martijnwalraven replied:
It is called [`Type`](http://spec.graphql.org/draft/#Type) in the spec, and that also lists `NamedType` as part of the recursive definition. Note that `NamedType` is not a replacement for `FieldValue` however, it is just one of the possible values of the enum (besides `NonNull` and `List`).

martijnwalraven replied:
If you don't like `Type`, maybe `TypeRef` would be an alternative?

lrlna replied:
Ok, we will have it as `Type_`, since `Type` is reserved.

martijnwalraven commented on crates/sdl-encoder/src/field_value.rs:
```diff
@@ -0,0 +1,120 @@
+    /// The non-null field type.
+    NonNull {
+        /// Null inner type.
+        ty: Box<FieldValue>,
```
Maybe `ofType`?

lrlna replied:
It's more clear to use `ty` as other encoders do.

martijnwalraven replied:
It is not just a type though, it is a `NonNull` or `List` of a specific type.

martijnwalraven commented on crates/sdl-encoder/src/field_value.rs:
```diff
@@ -0,0 +1,120 @@
+    /// The list field type.
+    List {
+        /// List inner type.
+        ty: Box<FieldValue>,
```
Maybe `ofType`?

martijnwalraven commented on crates/sdl-encoder/src/field_value.rs:
```diff
@@ -0,0 +1,120 @@
+    /// The type field type.
+    Type {
+        /// Type type.
+        ty: String,
```
Maybe `name`?

martijnwalraven commented on crates/sdl-encoder/src/input_object_def.rs:
```diff
@@ -0,0 +1,219 @@
+    // The vector of interfaces that this object implements.
+    interfaces: Vec<String>,
+    // A vector of fields
+    fields: Vec<Field>,
```
What is the reason for accepting `Field` here instead of `InputValue` or `InputField`? I saw your comment on the PR, but I don't really understand it:

> The exception is the `__InputObject` using the prescribed by the spec `__InputValue` type. As we do not have an AST to visit to encode the schema, all types that have fields currently use `__Field` type as their input.

Since we know when a type definition is an input object by looking at the `kind` field, and every input object always has `inputFields` but not `fields`, why can't this be reflected here?

lrlna replied:
Ok, we now have specific `InputField` encodings.

martijnwalraven commented on crates/sdl-encoder/src/input_object_def.rs:
```diff
@@ -0,0 +1,219 @@
+    // Description may return a String or null.
+    description: Option<String>,
+    // The vector of interfaces that this object implements.
+    interfaces: Vec<String>,
```
Input objects cannot implement interfaces.

martijnwalraven commented on crates/rover-client/src/introspection/schema.rs:
```diff
@@ -0,0 +1,237 @@
+                        input_def.interface(interface.type_ref.name.unwrap_or_else(String::new));
+                    }
+                }
+                if let Some(field) = ty.fields {
```
I think this should be `ty.input_fields`, since `fields` and `inputFields` are separate fields on `__Type`?

lrlna replied:
Ohhhhh, I've been thinking this whole time that I only get `fields` for this kind. The compiler never complained about the field `fields` not being present on `ty`, which I guess means that I get both.

martijnwalraven replied:
Yeah, in the introspection query most fields on `__Type` are optional since it describes all possible kinds of types.

lrlna replied:
It makes more sense to me now to have InputField and regular fields separated, like you've mentioned several times. It was weird to have essentially duplicate code to generate the exact same structure for the exact same return value of `ty.fields` before. Which is why I said that would make a lot more sense once we have an AST to walk and avoid having to write out the exact same code over the exact same set of values. 

I will still not be able to have both `InputObject` use `InputValue` since the formatting for an argument is different than a field:
```js
//  input values are single lined at the moment
name: String, val: Int, val2: Int
```
```js
// we will call these input fields so we can multiline them and format them similarly to fields
name: String
val: Int
val2: Int
```

It _is_ a bit weird to me that graphql spec does not differentiate between an input field and an input value, as they do format differently, but I guess that's not something to worry about atm

martijnwalraven commented on crates/rover-client/src/query/graph/introspect.rs:
```diff
@@ -0,0 +1,130 @@
+#[derive(Debug, PartialEq)]
+pub struct IntrospectionResponse {
+    pub result: String,
+}
```
Not sure I understand where this comes form or where it is used? Why is this about the gateway config?

lrlna replied:
the comment is old, sorry about that. 

`IntrospectionResponse` is used in this module as the response type we sent throughout to Rover. It's ends be being the encoded schema result we get back from the encoder.

martijnwalraven commented on crates/sdl-encoder/src/enum_def.rs:
```diff
@@ -0,0 +1,143 @@
+///   """Top bunk of a cat tree."""
+///   CatTree
+///   Bed
+///   CardboardBox @deprecated(reason: "Box was recycled.")
```
This is a small nit, but by common convention enum values in GraphQL are uppercased.

martijnwalraven commented on crates/sdl-encoder/src/directive_def.rs:
```diff
@@ -0,0 +1,170 @@
+impl Display for Directive {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(description) = &self.description {
+            // We are determing on whether to have description formatted as
+            // a multiline comment based on whether or not it already includes a
+            // \n.
+            match description.contains('\n') {
+                true => writeln!(f, "\"\"\"\n{}\n\"\"\"", description)?,
+                false => writeln!(f, "\"\"\"{}\"\"\"", description)?,
+            }
```
Maybe extract the multiline formatting code into a separate function to avoid code duplication?

martijnwalraven commented on crates/sdl-encoder/src/enum_value.rs:
```diff
@@ -0,0 +1,122 @@
+
+        write!(f, "  {}", self.name)?;
+
+        if self.is_deprecated {
+            write!(f, " @deprecated")?;
+            // Just in case deprecated directive is ever used without a reason,
+            // let's properly unwrap this Option.
+            if let Some(reason) = &self.deprecation_reason {
+                write!(f, "(reason: \"{}\")", reason)?;
+            }
+        }
```
Maybe the logic for formatting `@deprecated` could be extracted and reused in other places?

martijnwalraven commented on crates/sdl-encoder/src/field.rs:
```diff
@@ -0,0 +1,241 @@
+    // Deprecation reason optionally provides a reason why this field is deprecated.
+    deprecation_reason: Option<String>,
+    // Default is used to set a default value for fields **only** for fields in Input Objects.
+    default: Option<String>,
```
This shouldn't be necessary if we can use `InputValue` or `InputField` for fields of input objects.

martijnwalraven commented on crates/sdl-encoder/src/input_value.rs:
```diff
@@ -0,0 +1,183 @@
+    // the default value used by this input value in the condition a value is
+    // not provided at runtime. If this input value has no default value,
+    // returns null.
+    default: Option<String>,
```
Maybe `default_value`?

martijnwalraven commented on crates/sdl-encoder/src/input_value.rs:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    #[test]
+    fn it_encodes_valueument_with_deprecation() {
```
`valueument` might be a search/replace error?

martijnwalraven commented on crates/sdl-encoder/src/schema_def.rs:
```diff
@@ -0,0 +1,112 @@
+    description: Option<String>,
+    // The vector of fields in a schema to represent root operation type
+    // definition.
+    fields: Vec<Field>,
```
Since these aren't real fields (they can't have arguments) but references to operation root types, it might make more sense to include these as `query_type`, `mutation_type`, and `subscription_type` (as is done in `graphql-js` and other libraries).

lrlna replied:
Ah, ok!

Am I meant to encode the schema definition then based on the following values? 
```json
            "queryType": {
                "name": "Query"
            },
            "mutationType": null,
            "subscriptionType": null,
```

martijnwalraven replied:
Yep, the introspection query only asks for the `name`, which should be enough (there is no need for spreading a `TypeRef` fragment here because root types are always object types, never wrapped in `List` or `NonNull`).

lrlna replied:
oh thank god 🙏 at least one of these is straightforward

martijnwalraven commented on crates/sdl-encoder/src/union_def.rs:
```diff
@@ -0,0 +1,106 @@
+/// assert_eq!(
+///     union_.to_string(),
+///     r#""""A union of all cats represented within a household."""
+/// union Cat = NORI | CHASHU
```
Note that unions do not contain values as members, but always refer to object types.

lrlna replied:
I do not understand this comment. Do you mean the naming is incorrect? The [spec seems to call these values as `members`](https://spec.graphql.org/draft/#sec-Unions)

martijnwalraven replied:
What I mean is that `NORI` and `CHASHU` seem to be values (similar to enum values), but unions in GraphQL only contain object types as members. So this would rather be something like `union Pet = Cat | Dog`.

martijnwalraven commented on crates/sdl-encoder/src/union_def.rs:
```diff
@@ -0,0 +1,106 @@
+    // Description may return a String.
+    description: Option<String>,
+    // The vector of members that can be represented within this union.
+    members: Vec<String>,
```
It might be fine to represent these as strings for now if all we use it for is SDL output, but they really are references to object types, so these are type names.

lrlna replied:
This is an encoder, so it will only ever handle strings to generate a string representation of a schema. It is not the job of a serialiser/encoder to keep references to the object type, that is done in the AST.

martijnwalraven replied:
Ah, I think that is why some of my suggestions may not make as much sense to you. I was thinking of this as a schema definition layer similar to `GraphQLSchema` in `graphql-js`. We still need that if we want to do anything beyond encoding, but I can see how that isn't a priority right now. 

Note that the AST doesn't keep references to types either, it just contains names as strings (and also fails to encode other meaningful relationships). So I do think we need a layer of schema and type definitions as part of our `apollo-rs` library. At that point, it seems the types in `sdl-encoder` would be superfluous and maybe we could find other ways to add encoding behavior to a schema definition layer defined elsewhere (by implementing encoding traits for those types in `sdl-encoder` for example).

lrlna replied:
We will eventually have a schema definition layer you're talking about based on what's in `crates/rover-client/introspection/schema.rs`. Meaning it will encode its internal representation of the schema the way it does currently when we need an SDL. The current encoder is still missing some things to be able to do non-introspection based SDL serialisations, but that's something we will add later.

In general, I think discussing architecture of the future graphql rust tooling is out of scope for this PR review. I'd prefer to have that discussion separately as we carry out the planning and design steps for it.

lrlna committed:
chore: align naming conventions to the spec

Make sure encoder types are aligned to the spec:
- InputObjects now have InputFields instead of using Fields
- FieldValue is renamed to Type_ to align with graphql spec
- Type_'s named type is now referred to as `NamedType`

JakeDawkins committed:
feat(introspection): add header flag parser (#351)

Add the ability to pass through headers, similarly to how curl works with `--header` or `-H`.

```console
# single headers
rover graph introspect http://localhost:4000 --header auth:hello
rover graph introspect http://localhost:4000 -H auth:hello
rover graph introspect http://localhost:4000 -H "auth:hello"
rover graph introspect http://localhost:4000 --header auth:"hello"
rover graph introspect http://localhost:4000 --header "auth:value with spaces"
rover graph introspect http://localhost:4000 --header auth:"value with spaces"

# I think reqwest is stripping leading whitespace from header values automatically (which is what we'd want), 
# even though with the current implementation, leading whitespaces are added to the header map
rover graph introspect http://localhost:4000 --header "auth: value with leading spaces"

# multiple headers
rover graph introspect http://localhost:4000 --header auth:hello another:header
rover graph introspect http://localhost:4000 --header auth:hello --header another:header
rover graph introspect http://localhost:4000 --header auth:hello -H another:header
rover graph introspect http://localhost:4000 --header auth:hello -H "another-header:with spaces"
```

lrlna committed:
chore: introspection encoding integration tests

lrlna committed:
feat(sdl-encoder): encode schema definition

Schema Definition has three possible fields: query, mutation and
subscription. Create those fields directly with schema_def.query(),
schema_def.mutation() etc.

We encode Schema Definition depending on whether or not we have
mutations/subscriptions. If either of those is present, record all
possible values if there are any.

If neither mutation or subscription is present, record query if its
value is something other than `Query`.

lrlna committed:
Merge branch 'main' into build_schema

EverlastingBugstopper reviewed (approved):
✔️
